tools/explain_automaton.py

```
@@ -0,0 +1,155 @@
+            stderr=subprocess.PIPE,
+            text=True,
+            check=False,
+        )
+    if completed.returncode != 0:
+        sys.exit(f"menhir failed to preprocess {grammar}:\n{completed.stderr}")
+
+    base = directory / "automaton"
+    # A grammar with conflicts makes Menhir exit non-zero while still writing
+    # both files, and a grammar with conflicts is the whole subject here, so
+    # the status is ignored and the files are checked instead.
+    subprocess.run(
```

Ruff (0.16.1)

[error] `subprocess` call: check for execution of untrusted input

(S603)